src/bundles/game/functions.ts

```diff
@@ -15,773 +15,240 @@
-      get_game_obj(obj)
-        .setRotation(rad);
-      return obj;
-    }
-    throw_error(`${obj} is not of type ${ObjTypes}`);
-  }
+// List processing
+// Original Author: Martin Henz
 
-  function set_flip(
-    obj: GameObject,
-    x: boolean,
-    y: boolean,
-  ): GameObject | undefined {
-    const GameElementType = [ObjectTypes.ImageType, ObjectTypes.TextType];
-    if (is_any_type(obj, GameElementType)) {
-      (get_obj(obj) as RawGameElement).setFlip(x, y);
-      return obj;
-    }
-    throw_error(`${obj} is not of type ${GameElementType}`);
+/**
+ * array test works differently for Rhino and
+ * the Firefox environment (especially Web Console)
+ */
+function array_test(x: any): boolean {
+  if (Array.isArray === undefined) {
+    return x instanceof Array;
   }
+  return Array.isArray(x);
+}
 
-  async function add_tween(
-    obj: GameObject,
-    config: ObjectConfig = {},
-  ): Promise<GameObject | undefined> {
-    if (is_any_type(obj, ObjTypes)) {
-      scene.tweens.add({
-        targets: get_game_obj(obj),
-        ...config,
-      });
-      return obj;
-    }
-    throw_error(`${obj} is not of type ${ObjTypes}`);
-  }
+/**
+ * pair constructs a pair using a two-element array
+ * LOW-LEVEL FUNCTION, NOT SOURCE
+ */
+function pair(x: any, xs: any): [any, any] {
+  return [x, xs];
+}
 
-  // LISTENER
-
-  function add_listener(
-    obj: GameObject,
-    event: string,
-    callback: Function,
-  ): GameObject | undefined {
-    if (is_any_type(obj, ObjTypes)) {
-      const listener = get_game_obj(obj)
-        .addListener(event, callback);
-      return set_type(listener, ListenerTypes.InputPlugin);
-    }
-    throw_error(`${obj} is not of type ${ObjTypes}`);
-  }
+/**
+ * is_pair returns true iff arg is a two-element array
+ * LOW-LEVEL FUNCTION, NOT SOURCE
+ */
+function is_pair(x: any): boolean {
+  return array_test(x) && x.length === 2;
+}
 
-  function add_keyboard_listener(
-    key: string | number,
-    event: string,
-    callback: Function,
-  ): GameObject {
-    const keyObj = scene.input.keyboard.addKey(key);
-    const keyboardListener = keyObj.addListener(event, callback);
-    return set_type(keyboardListener, ListenerTypes.KeyboardKeyType);
+/**
+ * head returns the first component of the given pair,
+ * throws an exception if the argument is not a pair
+ * LOW-LEVEL FUNCTION, NOT SOURCE
+ */
+function head(xs: List): any {
+  if (is_pair(xs)) {
+    return xs![0];
   }
+  throw new Error(
+    `head(xs) expects a pair as argument xs, but encountered ${xs}`,
+  );
+}
 
-  function remove_listener(listener: GameObject): boolean {
-    if (is_any_type(listener, ListnerTypes)) {
-      get_input_obj(listener)
-        .removeAllListeners();
-      return true;
-    }
-    return false;
+/**
+ *  tail returns the second component of the given pair
+ * throws an exception if the argument is not a pair
+ * LOW-LEVEL FUNCTION, NOT SOURCE
+ */
+function tail(xs: List) {

```

Comment:
If you don't need any special functionality, would it be possible to import these list functions from the `js-slang` stdlib?

Reply:
Updated! Bumping for a review